### ISSUE TITLE
Strange warning from `matmul(..., out=...)`

### ISSUE THREAD

**Issue by thomasahle** [pytorch/pytorch#76779]
### 🐛 Describe the bug

The torch.matmul `out=` parameter seems to reshape its parameter down to two dimensions, which leads to a strange UserWarning:
```
U = torch.randn(2, 4, 3)
W = torch.randn(3, 5)
temp = torch.empty((U @ W).shape)
torch.matmul(U, W, out=temp)
UserWarning: An output with one or more elements was resized since it had shape [2, 4, 5], which does not match the required output shape [8, 5].This behavior is deprecated, and in a future PyTorch release outputs will not be resized unless they have zero elements. You can explicitly reuse an out tensor t by resizing it, inplace, to zero elements with t.resize_(0). (Triggered internally at  /Users/distiller/project/pytorch/aten/src/ATen/native/Resize.cpp:24.)
  torch.matmul(U, W, out=temp)
```
It seems like `temp` tensor should clearly have the right shape [2, 4, 5] and not [8, 5] like the warning says.

### Versions

Collecting environment information...
PyTorch version: 1.11.0
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 12.3.1 (x86_64)
GCC version: Could not collect
Clang version: 13.1.6 (clang-1316.0.21.2.3)
CMake version: version 3.21.4
Libc version: N/A

Python version: 3.9.12 (main, Apr 11 2022, 01:47:47)  [Clang 13.1.6 (clang-1316.0.21.2)] (64-bit runtime)
Python platform: macOS-12.3.1-x86_64-i386-64bit
Is CUDA available: False
CUDA runtime version: No CUDA
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

Versions of relevant libraries:
[pip3] functorch==0.1.1
[pip3] numpy==1.19.5
[pip3] torch==1.11.0
[pip3] torchvision==0.12.0
[conda] Could not collect

**Comment by bdhirsh** [comment]
I can repro on tip-of-trunk. We're probably doing some intermediate reshaping somewhere in the `matmul` kernel.

**Comment by ngimel** [comment]
Yeah there's a lot of intermediate reshaping going on in matmul. One case was fixed recently, but not this one.

**Comment by thomasahle** [comment]
If we start with an empty array, it works fine:
```
U = torch.randn(2, 4, 3)
W = torch.randn(3, 5)
temp = torch.empty(0)
torch.matmul(U, W, out=temp)
```
but if we then use the same tensor again, we run into trouble:
```
U = torch.randn(2, 4, 3)
W = torch.randn(3, 5)
temp = torch.empty(0)
torch.matmul(U, W, out=temp)
torch.matmul(U, W, out=temp)
UserWarning: ...
```

**Comment by thomasahle** [comment]
I guess an easy temporary workaround is just doing `temp.resize_(0)` before calling `torch.matmul`.
That doesn't release any memory, right?

**Comment by thomasahle** [comment]
Maybe the right behavior would be that no warnings are given as long as the total size (product) of the shapes match?

**Comment by ngimel** [comment]
No, underlying `resize_` function behavior is pretty confusing and it produces predictable results only if sizes match exactly. So we should fix this bug in matmul itself, it should not be that difficult.

**Comment by Lezcano** [comment]
This is fixed in https://github.com/pytorch/pytorch/pull/75197
